actionpack/lib/action_controller/metal/request_forgery_protection.rb

```diff
@@ -460,16 +552,63 @@ def non_xhr_javascript_response? # :doc:
 
       AUTHENTICITY_TOKEN_LENGTH = 32
 
-      # Returns true or false if a request is verified. Checks:
+      # Safe values for Sec-Fetch-Site header that indicate the request
+      # originated from the same site.
+      SAFE_FETCH_SITES = %w[ same-origin same-site ].freeze
+      private_constant :SAFE_FETCH_SITES
+
+      # Returns true or false if a request is verified. The verification method
+      # depends on the configured `forgery_protection_verification_strategy`:
+      #
+      # *   `:fetch_metadata` - Uses Sec-Fetch-Site header only (default)
+      # *   `:token_fallback` - Uses Sec-Fetch-Site header with fallback to token
       #
-      # *   Is it a GET or HEAD request? GETs should be safe and idempotent
-      # *   Does the form_authenticity_token match the given token value from the
-      #     params?
-      # *   Does the `X-CSRF-Token` header match the form_authenticity_token?
+      # For all strategies, GET and HEAD requests are allowed without verification.
       #
       def verified_request? # :doc:
+        return true if request.get? || request.head? || !protect_against_forgery?
+
         request.get? || request.head? || !protect_against_forgery? ||
-          (valid_request_origin? && any_authenticity_token_valid?)
+          (valid_request_origin? && verified_request_for_forgery_protection?)
+      end
+
+      def verified_request_for_forgery_protection?
+        if using_only_fetch_metadata_for_forgery_protection?
+          verified_via_fetch_metadata?
+        else
+          verified_with_token_fallback?
+        end
+      end
+
+      def using_only_fetch_metadata_for_forgery_protection?
+        forgery_protection_verification_strategy == :fetch_metadata
+      end
+
+      def verified_via_fetch_metadata?
+        SAFE_FETCH_SITES.include?(sec_fetch_site_value) ||
+          (sec_fetch_site_value == "cross-site" && origin_trusted?)
+      end
+
+      def verified_with_token_fallback?
+        case sec_fetch_site_value
+        when "same-origin", "same-site"
+          true
+        when "cross-site"
+          origin_trusted?
+        else # "none" or missing
+          logger.warn "Falling back to CSRF token check for forgery protection" if logger && log_warning_on_csrf_failure
```

Comment:
IDK if this should be Rubocop rule or documented somewhere, but I think Rails framework components shouldn't log directly and use Active Support Notifications (or the new Event Reporter) instead so that applications can have control over the log output and hook up their own additional behaviour in a clean way.

Reply:
Good point! I had used Active Support notifications on my first approach, but changed to this to match the other logging done in this module. Maybe everything should be switched instead of mixing the two ways?

Reply:
Agree re. sticking with existing usage. We can introduce structured events in a separate PR.